cpp/include/cudf/column/column.hpp

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/types.hpp>
+#include "column_view.hpp"
+
+#include <rmm/device_buffer.hpp>
+#include <rmm/mr/device_memory_resource.hpp>
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace cudf {
+
+class column {
+ public:
+  ~column() = default;
+  column& operator=(column const& other) = delete;
+  column& operator=(column&& other) = delete;
+  column() = default;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Construct a new column by deep copying the device memory of another
+   * column.
+   *
+   * All device memory allocation and copying is done using the
+   * `device_memory_resource` and `stream` from `other`.
+   *
+   * @param other The other column to copy
+   *---------------------------------------------------------------------------**/
+  column(column const& other);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Construct a new column object from the contents of another `column`
+   * using move semantics. After the move, `other` is guaranteed to be empty.
+   *
+   * @param other The other column whose device memory will be moved to the new
+   * column
+   *---------------------------------------------------------------------------**/
+  column(column&& other);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Construct a new column
+   *
+   * @param[in] dtype The element type
+   * @param[in] size The number of elements in the column
+   * @param[in] data `device_buffer` whose contents will be copied for the
```

Comment:
Must it be a `device_buffer`? Could it copy from an arbitrary device pointer? A `device_vector`?